### PR TITLE
add StorageRedis and TableFunctionRedis

### PR DESCRIPTION
**Pull request by aaapetrenko**
### Changelog category (leave one):
- New Feature

### CODE REVIEW

File: src/Storages/StorageRedis.cpp [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+        auto configuration = StorageRedis::getConfiguration(args.engine_args, args.getLocalContext());
+        std::string primary_key;
+        std::string secondary_key;
+        std::string value;
+        if (configuration.storage_type == StorageRedisConfiguration::RedisStorageType::HASH_MAP) {
+            primary_key = args.columns.getNamesOfPhysical()[0];
+            secondary_key = args.columns.getNamesOfPhysical()[1];
+            value = args.columns.getNamesOfPhysical()[2];
+        } else {
+            secondary_key = args.columns.getNamesOfPhysical()[0];
```
Comment (CurtizJ) [inline src/Storages/StorageRedis.cpp:220]: Why not primary key?

File: src/Storages/StorageRedis.cpp [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+        {
+            if (client->isConnected())
+                client->disconnect();
+
+            pool->returnObject(std::move(client));
+            throw;
+        }
+    }
+    return std::make_unique<Connection>(pool, std::move(client));
+}
```
Comment (CurtizJ) [inline src/Storages/StorageRedis.cpp:290]: Also copy-paste from Redis dictionary.

File: src/TableFunctions/TableFunctionMongoDB.cpp [hunk 3]
```diff
@@ -0,0 +1,104 @@
+#include <Common/Exception.h>
```
Comment (CurtizJ) [inline src/TableFunctions/TableFunctionMongoDB.cpp:1]: Accidentally committed from #37213?

File: src/Storages/StorageRedis.cpp [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+StorageRedis::StorageRedis(
+        const StorageID & table_id_,
+        const std::string & host_,
+        const UInt16 & port_,
+        const UInt32 & db_index_,
+        const std::string & password_,
+        const RedisStorageType & storage_type_,
+        const String & primary_key_column_name_,
+        const String & secondary_key_column_name_,
+        const String & value_column_name_,
```
Comment (CurtizJ) [inline src/Storages/StorageRedis.cpp:40]: Why do we need those names as separate fields? They are already stored in `ColumnsDescription & columns_`?

File: src/TableFunctions/TableFunctionRedis.cpp [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+    String{});
+    storage->startup();
+    return storage;
+}
+
+ColumnsDescription TableFunctionRedis::getActualTableStructure(ContextPtr context) const
+{
+    String structure;
+    if (configuration->storage_type ==  StorageRedisConfiguration::RedisStorageType::HASH_MAP)
+        structure = "primary_key String, secondary_key String, value String";
```
Comment (CurtizJ) [inline src/TableFunctions/TableFunctionRedis.cpp:50]: Maybe use terminology from Redis documentation: https://redis.io/commands/hget/ and use column names: `key`, `field`, `value`?

**Review by CurtizJ — COMMENTED** [review]
Please, add tests for Redis storage. I see only tests for MongoDB.

**Comment by alexey-milovidov** [comment]
Abandoned.

**Pull request closed without merging**